### PR TITLE
Add support for locking and unlocking Rules

### PR DESCRIPTION
**Pull request by Bialogs**
* Create Rules table with lock column
* Add an admin column to User to facilitate basic unlocking
* Add callback to Rule to prevent further editing if lock is set
* Add Rule#unlock to allow unlocking Rules when permissions are correct
* Add User#can_unlock to manage permissions around unlocking Rules

### CODE REVIEW

File: app/models/rule.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class Rule < ApplicationRecord
+  before_validation :error_if_locked, on: :update
```
Comment (jkufro) [inline app/models/rule.rb:4]: I'm wondering if we should also protect locked rules from being deleted

File: app/models/user.rb [hunk 2]
```diff
@@ -22,4 +22,8 @@ def self.from_omniauth(auth)
       user.skip_confirmation!
     end
   end
+
+  def can_unlock?(_rule)
```
Comment (jkufro) [inline app/models/user.rb:26]: Will the roles able to unlock be different than those that can lock? If not I think this method would be better named as `can_manage_rule_lock?` and be used for checking permissions on lock and unlock

File: app/models/rule.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+  # Allow an authorized user to unlock a rule
+  def self.unlock(rule, user)
+    raise(RuleLockedError, rule.id) unless user.can_unlock?(rule)
```
Comment (jkufro) [inline app/models/rule.rb:8]: 1. Where will privilege checking for locking a rule be done?
2. Is the idea to have an `update` and a `unlock` controller action in the RuleController?

**Review by jkufro — CHANGES_REQUESTED** [review]
Se inline comments/questions about rules

File: app/models/rule.rb (hunk 3 above)
Comment (jkufro) [reply app/models/rule.rb:8]: Could this be more elegant and set up for authorization in future actions if we remove this static `self.unlock` method and write the manage lock controller action now with pundit?

https://github.com/varvet/pundit

```ruby
class RulesController < ApplicationController
  def update
    authorize @rule
    # TODO
  end

  def manage_lock
    authorize @rule
    rule.locked = manage_lock_params[:rule][:locked]
    rule.save(validate: false)
  end

  private

  def update_params
    params.require(:rule).permit(:todo)
  end

  def manage_lock_params
    params.require(:rule).permit(:locked)
  end
end

class RulePolicy
  attr_reader :user, :rule

  def initialize(user, rule)
    @user = user
    @rule = rule
  end

  def update?
    # TODO
    true
  end

  def manage_lock?
    user.can_unlock?(rule)
  end
end
```

Comment (Bialogs) [reply app/models/rule.rb:9]: Yes we would want a controller action for lock and update

Comment (jkufro) [reply app/models/rule.rb:9]: What is the reasoning for separating lock/unlock into two separate controller actions?

**Pull request merged**